pkg/kubelet/kubelet.go

```diff
@@ -2555,6 +2564,136 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+	containerName string,
+	options *runtimeapi.CheckpointContainerRequest,
+	checkpointTime int64,
+) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// set this to 10 as a default value if 0 or not set. The error
+		// handling is mainly here to make sure tests are working.
+		return fmt.Errorf("maxContainerCheckpoints needs to > 0")
+	}
+
+	type checkpoint struct {
+		Archive   string `json:"archive,omitempty"`
+		Size      int64  `json:"size,omitempty"`
+		Timestamp int64  `json:"timestamp,omitempty"`
+	}
+	type containerCheckpointMetadata struct {
+		PodFullName   string       `json:"podFullName,omitempty"`
+		ContainerName string       `json:"containerName,omitempty"`
+		TotalSize     int64        `json:"totalSize,omitempty"`
+		Checkpoints   []checkpoint `json:"checkpoints"`
+	}
+
+	checkpointJsonName := fmt.Sprintf(
+		"checkpoint-%s-%s.json",
+		podFullName,
+		containerName,
+	)
+
+	checkpointJsonPath := filepath.Join(kl.getCheckpointsDir(), checkpointJsonName)
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.updateMaxContainerCheckpoints.Lock()
+	defer kl.updateMaxContainerCheckpoints.Unlock()
+
+	checkpointMetadata, err := func() (*containerCheckpointMetadata, error) {
+		_, err := os.Stat(checkpointJsonPath)
+		// If the file does not exist return an empty structure.
+		if err != nil {
+			return &containerCheckpointMetadata{}, nil
+		}
+		// If the file does exist return that content.
+		content, err := os.ReadFile(checkpointJsonPath)
+		if err != nil {
+			return nil, err
+		}
+		ccm := &containerCheckpointMetadata{}
+		if err = json.Unmarshal(content, ccm); err != nil {
+			return nil, err
+		}
+		return ccm, nil
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	fi, err := os.Stat(options.Location)
+	if err != nil {
+		return err
+	}
+
+	chkpts := &checkpointMetadata.Checkpoints
+	numberOfCheckpoints := len(*chkpts)
+
+	for numberOfCheckpoints >= kl.maxContainerCheckpoints {
+		klog.InfoS(
+			"More than maxContainerCheckpoints",
+			"container", containerName,
+			"numberOfCheckpoints", numberOfCheckpoints,
+			"maxContainerCheckpoints", kl.maxContainerCheckpoints,
+		)
+		// Find oldest checkpoint and delete it.
+		oldest := (*chkpts)[0].Timestamp
+		entryToDelete := 0
+		trimSizeBy := int64(0)
+		for i, checkpoint := range *chkpts {
+			if checkpoint.Timestamp < oldest {
+				oldest = checkpoint.Timestamp
+				entryToDelete = i
+				trimSizeBy = checkpoint.Size
+			}
+		}
+		klog.InfoS(
+			"Deleting container checkpoint archive",
+			"container", containerName,
+			"archive", (*chkpts)[entryToDelete].Archive,
+		)
+		// Ignoring error during delete in case the checkpoint archive
+		// has already been deleted.
+		_ = os.Remove((*chkpts)[entryToDelete].Archive)
+		// Also remove from slice.
+		(*chkpts)[entryToDelete] = (*chkpts)[len(*chkpts)-1]
+		(*chkpts) = (*chkpts)[:len(*chkpts)-1]
```

Comment:
This is probably a case of premature optimization from my side. The code is indeed not really good to read. I just wanted to avoid unnecessary memory usage. I removed the pointer which I just introduced to have a shorter identifier. Now the lines are really long. Let me know if you would prefer some other solution to modify the slice.